File: ksh/week5/조이스틱.java
package week5;

public class 조이스틱 {
    public int solution(String name) {
        int count = 0;
        int len = name.length();
        int move = len - 1; // 오른쪽으로만 이동했을 때의 기본 이동 횟수

        for (int i = 0; i < len; i++) {
            count += find(name.charAt(i));

            // 다음 알파벳으로 이동
            int next = i + 1;
            while (next < len && name.charAt(next) == 'A') {
                next++;
            }

            // 왼쪽으로 이동하는 경우와 오른쪽으로 이동하는 경우 중 더 작은 값을 선택
            move = Math.min(move, i + len - next + Math.min(i, len - next));
        }

        count += move;

        return count;
    }

    public int find(char ch) {
        if(ch <= 77)
            return ch - 'A';
        return 1 + 'Z' - ch;
    }
}

// 알파벳 개수 26개
// A B C D E F G H I J K L |M| |N| O P Q R S T U V W X Y Z
// M : 77
// 아스키 A : 65 / Z : 90

File: lys/조이스틱.js
function solution(name) {
    let totalMoves = 0; 
    const asciiA = 65; 
    const asciiZ = 90; 
    
    const getMove = (targetChar) => {
        const targetAscii = targetChar.charCodeAt();
        return Math.min(targetAscii - asciiA, asciiZ - targetAscii + 1);
    }

    const charMoves = [];
    for (let i = 0; i < name.length; i++) {
        const char = name[i];
        const move = getMove(char);
        charMoves.push(move);
    }
    
    let minTotalMove = charMoves.length - 1;
    for (let index = 0; index < charMoves.length; index++) {
        const move = charMoves[index];
        totalMoves += move;

        let nextIndex = index + 1;
        while (nextIndex < charMoves.length && charMoves[nextIndex] === 0) {
            nextIndex++;
        }

        minTotalMove = Math.min(minTotalMove, (index * 2) + charMoves.length - nextIndex, index + 2 * (charMoves.length - nextIndex));
    }

    return totalMoves + minTotalMove;
}
